libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/BannerInstructionEvent.kt

```diff
@@ -1,34 +1,58 @@
 package com.mapbox.navigation.core.trip.session
 
 import com.mapbox.api.directions.v5.models.BannerInstructions
+import com.mapbox.navigation.utils.internal.ifNonNull
 
-internal class BannerInstructionEvent {
+internal class BannerInstructionEvent private constructor() {
 
-    var latestInstructionIndex: Int? = null
+    var latestInstructionWrapper: LatestInstructionWrapper? = null
         private set
 
+    val latestInstructionIndex: Int?
+        get() = latestInstructionWrapper?.latestInstructionIndex
+
+    val latestBannerInstructions: BannerInstructions?
+        get() = latestInstructionWrapper?.latestBannerInstructions
+
     var bannerInstructions: BannerInstructions? = null
         private set
 
-    var latestBannerInstructions: BannerInstructions? = null
-        private set
+    companion object {
+        operator fun invoke(): BannerInstructionEvent = BannerInstructionEvent()
+    }
 
     fun isOccurring(bannerInstructions: BannerInstructions?, instructionIndex: Int?): Boolean {
         return updateCurrentBanner(bannerInstructions, instructionIndex)
     }
 
-    fun invalidateLatestBannerInstructions() {
-        latestBannerInstructions = null
-        latestInstructionIndex = null
+    fun invalidateLatestBannerInstructions(latestInstructionWrapper: LatestInstructionWrapper?) {
+        if (latestInstructionWrapper == this.latestInstructionWrapper) {
```

Comment:
Shouldn't you synchronize (use mutex) here? What if `this.latestInstructionWrapper` changes between the lines 29 and 30? It can be done from `updateCurrentBanner` which can probably be executed from a different coroutine.

Reply:
the was in `navigator.updateLegIndex(legIndex)` where we delegate thread work into NN, when the control has returned to the SDK it's not required to be synchronized. Also, mutex and other concurrency stuff add CPU cost. Let's avoid using it where it's possible

Reply:
I still don't see how we can guarantee that line 37 will not be executed between lines 29 and 30 from a different coroutine.

Reply:
discussed: the coroutines are not interrupted here and no need for concurrency synchronization mechanisms